File: 1768-merge-strings-alternately/1768-merge-strings-alternately.py
class Solution:
    def mergeAlternately(self, word1: str, word2: str) -> str:
        # TC: O(max(m, n)); SC: O(1) 
        res = []
        l1, l2 = len(word1), len(word2)
        overlap_len = min(l1, l2)

        for i in range(overlap_len):
            res.append(word1[i] + word2[i])
        
        if (l1 > l2):
            res.append(word1[overlap_len :])
        else:
            res.append(word2[overlap_len :])

        return "".join(res)

        

File: 1768-merge-strings-alternately/1768-merge-strings-alternately.scala
object Solution {
    def mergeAlternately(word1: String, word2: String): String = {
        // TC: O(max(m, n)); SC: O(1)
        val res = new StringBuilder
        val l1 = word1.length
        val l2 = word2.length
        val overlapLen = Math.min(l1, l2)

        for (i <- 0 until overlapLen) {
            res.append(word1(i))
            res.append(word2(i))
        }

        if (l1 > l2) {
            res.append(word1.substring(overlapLen))
        } else {
            res.append(word2.substring(overlapLen))
        }

        res.toString
        
    }
}